### PR TITLE
Refactor Mains to use common BitcoinSRunner

### LINKED ISSUE #2116: Refactor commonalties between Main's into a single trait
In #2105 we introduce another Main.scala. From what I can tell there is a lot of duplication between the 2. The configuration should be factored out so we don't need to remember to update it in two places.

### PR DESCRIPTION
**Pull request by benthecarman**
Fixes #2116

`BitcoinSServerMain` is a little funky so we can test it

### CODE REVIEW

File: app/server/src/main/scala/org/bitcoins/server/BitcoinSRunner.scala [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+      val str = args(configIndex + 1)
+      val usableStr = str.replace("~", Properties.userHome)
+      val path = Paths.get(usableStr)
+      ConfigFactory
+        .parseFile(path.toFile)
+        .resolve()
+  }
+
+  lazy val configDataDir: Path = Paths.get(
+    baseConfig.getStringOrElse("bitcoin-s.datadir",
```
Comment (Christewart) [inline app/server/src/main/scala/org/bitcoins/server/BitcoinSRunner.scala:72]: I don't have anything actionable on this for now, but we do want to make sure we are compatible with the resolution hiearchy in the type safe config specification. I listed it out on #2147

Comment (Christewart) [reply app/server/src/main/scala/org/bitcoins/server/BitcoinSRunner.scala:72]: Actually thinking about this more, is there a reason we don't move 

1. configDataDir
2. datadirPath
3. network 

Could these all be made available via a `BitcoinSAppConfig` in this trait?

File: app/server/src/main/scala/org/bitcoins/server/BitcoinSServerMain.scala [hunk 2]
```diff
@@ -256,12 +177,13 @@ object Main extends App with HttpLogger {
   }
 
   //start everything!
-  val run = runMain(args.toVector)
-
-  run.failed.foreach { err =>
-    logger.error(s"Failed to startup server!", err)
-    sys.exit(1)
-  }(scala.concurrent.ExecutionContext.Implicits.global)
+  lazy val run: Unit = {
```
Comment (Christewart) [inline app/server/src/main/scala/org/bitcoins/server/BitcoinSServerMain.scala:181]: Why is this a `lazy val` ? Maybe make this a `def run(): Unit` ?

File: app/server/src/main/scala/org/bitcoins/server/BitcoinSServerMain.scala [hunk 3]
```diff
@@ -27,91 +22,17 @@ import org.bitcoins.wallet.Wallet
...
-        case MainNet  => "mainnet"
-        case TestNet3 => "testnet3"
-        case RegTest  => "regtest"
-        case SigNet   => "signet"
-      }
-      datadirPath.resolve(lastDirname)
-    }
 
-    System.setProperty("bitcoins.log.location", datadir.toAbsolutePath.toString)
+class Main(override val args: Array[String]) extends BitcoinSRunner {
```
Comment (Christewart) [inline app/server/src/main/scala/org/bitcoins/server/BitcoinSServerMain.scala]: I would really rather not call this or the object below `Main` as it starts to get confusing when searching in intellij. Maybe call this `BitcoinSServerMain`?

Comment (Christewart) [reply app/server/src/main/scala/org/bitcoins/server/BitcoinSServerMain.scala]: you can always specify the `Main` that gets run in the build file here: https://github.com/bitcoin-s/bitcoin-s/blob/b1635a8f70b5a50cfd4c00c9781587cc0935127f/app/server/server.sbt#L9-L8

File: app/server/src/main/scala/org/bitcoins/server/BitcoinSServerMain.scala [hunk 4]
```diff
@@ -377,6 +298,10 @@ object Main extends App with HttpLogger {
   }
 }
 
+object Main extends App {
```
Comment (Christewart) [inline app/server/src/main/scala/org/bitcoins/server/BitcoinSServerMain.scala]: Like I said above, i would prefer to name this something else.

**Review by Christewart — CHANGES_REQUESTED** [review]

File: app/server/src/main/scala/org/bitcoins/server/BitcoinSServerMain.scala (hunk 2 above)
Comment (benthecarman) [reply app/server/src/main/scala/org/bitcoins/server/BitcoinSServerMain.scala:181]: Made it a lazy val so we can't accidentally start the server twice

File: app/server/src/main/scala/org/bitcoins/server/BitcoinSRunner.scala (hunk 1 above)
Comment (benthecarman) [reply app/server/src/main/scala/org/bitcoins/server/BitcoinSRunner.scala:72]: 1) We need all of these so we can properly set the log location.

2) We only use a `BitcoinSAppConfig` in the app server, the oracle server uses a `DLCOracleAppConfig`

**Review by Christewart — APPROVED** [review]

**Pull request merged**